pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,244 @@ var (
 	}
 )
 
-// Requirements is a decorated alias type for []v1.NodeSelectorRequirements
-type Requirements []v1.NodeSelectorRequirement
-
-func (r Requirements) Zones() sets.String {
-	return r.Requirement(v1.LabelTopologyZone)
-}
-
-func (r Requirements) InstanceTypes() sets.String {
-	return r.Requirement(v1.LabelInstanceTypeStable)
+type Requirements struct {
+	// Requirements are layered with Labels and applied to every node.
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	// Ground truth for record keeping
+	Records map[string][]v1.NodeSelectorRequirement `json:"-"`
+	Allows  map[string]sets.Set                     `json:"-"`
 }
 
-func (r Requirements) Architectures() sets.String {
-	return r.Requirement(v1.LabelArchStable)
+func NewRequirements(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	result := Requirements{
+		Records: map[string][]v1.NodeSelectorRequirement{},
+		Allows:  map[string]sets.Set{},
+	}
+	return result.Add(requirements...)
 }
 
-func (r Requirements) OperatingSystems() sets.String {
-	return r.Requirement(v1.LabelOSStable)
+func (r *Requirements) Add(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	if r == nil {
+		return NewRequirements(requirements...)
+	}
+	rd := r.DeepCopy()
+	for _, requirement := range requirements {
+		// Normalizing the key
+		if newKey, ok := NormalizedLabels[requirement.Key]; ok {
+			requirement.Key = newKey
+		}
+		//rd.Requirements = append(rd.Requirements, requirement)
```

Comment:
Don't forget about this guy.

Reply:
I am removing that line. I used it for debugging.